Support for hardware hashing in the Hotel chip

yuriks opened the pull request:
I still have a few questions about some things, which I've left as inline comments.

yuriks commented on apps/libs/digest.h:
```diff
@@ -0,0 +1,34 @@
+#include <stdlib.h>
+
+// TODO These need to be standardized
+#define HOTEL_DRIVER_DIGEST 2
```
I feel like we should be standardizing these across chips, because otherwise there's going to just be no way of writing crossplatform programs/examples.

alevy replied:
Yes I agree 100%. As it stands, since there is really only one userspace, it's kind of de-facto standardized, but that won't be the case for very long. (a) we should actually do it systematically, (b) we should write it down somewhere.

I would suggest having low numbered system call major numbers be for very common drivers, in decreasing order of ubiquity/utility, and reserve higher numbered major numbers for more platform specific drivers.

For example, a mapping could be something like (large numbers are represented below as negative numbers):

```
0 => LEDs
1 => GPIOs
2 => Timer
3 => Console
...
9 => RNG
10 => SHA256
...
-3 => Platform specific driver
-2 => Platform specific driver
-1 => Platform specific driver
``` 

Worth you suggesting like this to the list?

(LEDs and GPIO drivers above are very similar, or even identical, but it makes sense to separate for userspace so an app doesn't have to know which pins are connected to LEDs --- might make sense to have the same for buttons, but this is all a bit besides the general point)

yuriks replied:
Yeah, that's pretty much what I had in mind. I'd only move the platform drivers to something like 0x10000 instead, because negative number could be re-purposed to mean something else.

yuriks commented on apps/sha_test/main.c:
```diff
@@ -0,0 +1,95 @@
+        }
+}
+
+int main(void) {
```
TODO: Text explanation of what each of the blink patterns means (or just rebase on top of UART support and print to console)

alevy replied:
UART/console is merged now, so `printf` from userland works

yuriks commented on src/main/src/callback.rs:
```diff
@@ -1,7 +1,7 @@
 use process;
 
-#[derive(Clone,Copy)]
+#[derive(Clone,Copy,Eq,PartialEq)]
```
Not sure about this one. I use it to check if the application calling the functions is the one that "owns" the device, but is this identifier supposed to be stable over calls? (Works on my machine™, in any case.)

alevy replied:
This is totally fine

yuriks commented on src/drivers/src/digest.rs:
```diff
@@ -0,0 +1,140 @@
+                        .unwrap_or(Err(SyscallError::Unknown))
+                }
+                _ => Err(SyscallError::NotImplemented),
+            }
```
I'm not sure why rustfmt things it should double-indent the match body, or why it should leave this closing bracket here hanging.

yuriks commented on src/drivers/src/digest.rs:
```diff
@@ -0,0 +1,140 @@
+
+                            Ok(0)
+                        })
+                        .unwrap_or(Err(SyscallError::Unknown))
```
Should I just panic here instead? Under what situations could this happen that aren't a kernel bug?

alevy replied:
Like my comment regarding `TakeCell` in #22, a driver should basically no panic because of a bug (if it's avoidable). The way I think of it is that these drivers are not really part of the "kernel" (even though they run in protected mode).

Basically, this seems good to me :)

yuriks replied:
In that case I'll introduce another error code for those cases. I think that, rather than TakeCell we should use RefCell with try_borrow, because that's ergonomically better to me than the map with a lambda, which makes things like error returns with try! quite painful.

alevy replied:
I can't find the definition for try_borrow, but you can also use `TakeCell`'s `take` to get an `Option` (map uses this internally).

In general I think we ought to avoid constructs which in normal Rust generally panic. We can always extend TakeCell with more methods if ergonomics is the problem

alevy replied:
OK, I found `try_borrow`. That would be a fine thing to use except I still think we should avoid `RefCell` since it _can_ panic (I guess if it's really cumbersome to use, maybe RefCell with a "TODO" comment is fine assuming we'll have a method for TakeCell that returns a `Result`)

We should probably have this discussion on the tock-dev mailing list, but do you think, in general, we should be opting for `Result` vs `Option` types for these kinds of things? Specifically, do you have the sense that `try!` make much of this programming easier?

`Option` wasn't really an informed decision, just because I'm used to using it in Haskell (where Monads sugar solves this problem). But if Result is more cannonical Rust, we should switch to that.

yuriks replied:
It's not so much the use of `Option` (which can be readily converted into a `Result` using `ok_or`), but the idiom of using `map` with `TakeCell`. The lambda nesting means you end up with non-flat control flow and types like `Result<Option<Result<T, E1>>, E2>` which end being actually pretty painful to collapse down to the `Result<T, E2>` you want. (As an example. I did run into this issue before when using `TakeCell` but don't remember the exact type.)

There's nothing that makes this unsolvable, I just didn't consider the burden of introducing a new cell type when `RefCell` is already in core and effectively has essentially the same semantics a worthwhile trade-off. But you're right, this is probably be a discussion for the mailing list.

I've started fixing this code to remove the panics, just didn't have time to finish it up today.

yuriks commented:
Updated.

To get around the printing issue, I implemented a crude usermode workaround of only sending 48 bytes at a time to the kernel.

Turns out that `try_borrow` has only been added recently in August, so your nightly version doesn't have that. It's going to be stabilized, so I'm going to open a discussion in the list about using that instead: I think I just threw up in my mouth a little while writing that last commit... :)

I also added a new error code to cover these "impossible" states that shouldn't happen. I'll work on upstreaming the syscall error codes and standardizing the driver numbers eventually.

yuriks commented:
Travis seems to be failing on rustfmt, reporting that a bunch of blank lines need to be removed at the beginning of several files. Doesn't make much sense since I didn't touch a lot of those, and running the same command locally doesn't say anything.

alevy commented:
@yuriks yes, I've fixed this in master --- it's just because of a new version of rustfmt that gets installed by default. We can ignore.
